webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/batch/PurgeOldInventoryFiles.scala

```diff
@@ -0,0 +1,121 @@
+import com.normation.utils.CronParser._
+import com.normation.zio._
+import cron4s.CronExpr
+import java.time.Instant
+import zio._
+import zio.clock.Clock
+import zio.duration._
+
+/**
+ * A scheduler which delete old inventory files under /var/rudder/inventories/{received, failed}
+ * It use a cron-like config parsed by cron4s.
+ */
+class PurgeOldInventoryFiles(
+    optCron:          Option[CronExpr],
+    maxAge:           Duration,
+    cleanDirectories: List[File]
+) {
+
+  val logger = InventoryProcessingLogger
+
+  // check is a file is older than max age (true) or not. Only true for regular file.
+  // maxAge in seconds.
+  def isOlderThanMaxAge(f: File, maxAge: Long, now: Instant): IOResult[Boolean] = {
+    IOResult.effect(f.isRegularFile && f.attributes().creationTime().toInstant.isBefore(now.minusSeconds(maxAge)))
+  }
+
+  // must not fail, will be in a cron
+  val cleanOldFiles: UIO[Unit] = for {
+    t0 <- currentTimeMillis
+    now = Instant.ofEpochMilli(t0)
+    _  <- ZIO
+            .foreach_(cleanDirectories) { dir =>
+              ZIO.foreach_(dir.list.to(Iterable)) { f =>
+                isOlderThanMaxAge(f, maxAge.toSeconds, now).flatMap { older =>
+                  if (older) {
+                    InventoryProcessingLogger.info(
+                      s"Deleting inventory file '${f.pathAsString}' (older than ${maxAge.toString})"
+                    ) *>
+                    IOResult
+                      .effect(f.delete())
+                      .catchAll(err =>
+                        InventoryProcessingLogger.error(s"Error when trying to clean old inventory file '${f.pathAsString}'")
+                      )
+                  } else {
+                    InventoryProcessingLogger.trace(
+                      s"Keeping inventory file '${f.pathAsString}' (younger than ${maxAge.toString})"
+                    )
+                  }
+                }
+              }
+            }
+            .catchAll(err => logger.error(s"Error when cleaning-up old inventory files: ${err.fullMsg}"))
+
+    t1 <- currentTimeMillis
+    _  <- logger.debug(s"Cleaned-up old inventory files in ${Duration.fromMillis(t1 - t0).toString}")
+  } yield ()
+
+  // create the schedule cron or nothing if disabled.
+  // Must not fail.
+  val prog: URIO[Any with Clock, Unit] = optCron match {
+    case None       =>
+      logger.info(
+        s"Disable automatic cleaning of old inventories in ${cleanDirectories.map(_.pathAsString).mkString(", ")} (schedule: '${DISABLED}')"
+      )
+    case Some(cron) =>
+      val schedule = cron.toSchedule
+      logger.info(
+        s"Automatic cleaning of file older than ${maxAge.toString} starts for directories: '${cleanDirectories
+            .map(_.pathAsString)
+            .mkString("', '")}' (schedule 'sec min h dayMonth month DayWeek': '${cron.toString}')"
+      ) *>
+      cleanOldFiles.schedule(schedule).unit
+  }
+
+  // start cron
+  def start() = {
```

Comment:
what about returning the Fiber (or something closable) to let the caller handle the lifecycle (for example in case of tests)?

Reply:
It could be interesting. In general, I prefer to have the caller do the least possible things in the first implementations and open up API when needs arise. 
For that case, I think I just didn't think to the that possibility / way of doing it.